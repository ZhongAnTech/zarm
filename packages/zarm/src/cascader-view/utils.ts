import { FieldNames } from '../picker-view/interface';
import { resolvedFieldNames } from '../picker-view/utils';
import type {
  BaseCascaderViewProps,
  CascaderItem,
  CascaderOption,
  CascaderValue,
} from './interface';

export const parseState = (props: BaseCascaderViewProps) => {
  const { value, defaultValue } = props;
  const currentValue: CascaderValue[] = [];

  return {
    value: value || defaultValue,
    currentValue,
  };
};

export const parseItems = (
  options: CascaderOption[],
  value: CascaderValue[],
  fieldNames?: Partial<FieldNames>,
) => {
  const field = resolvedFieldNames(fieldNames);
  const items: CascaderItem[] = [];
  const traverse = (memo: CascaderOption[], depth = 0) => {
    memo.forEach((option) => {
      const children = option[field.children] as CascaderOption[] | undefined;
      const currentValue = value[depth];
      if (option[field.value] === currentValue) {
        items[depth] = option;

        if (Array.isArray(children) && children.length > 0) {
          traverse(children, depth + 1);
        }
      }
    });
  };

  traverse(options);

  return items;
};
