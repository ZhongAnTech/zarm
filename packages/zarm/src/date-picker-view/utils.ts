import dayjs from 'dayjs';
import isLeapYear from 'dayjs/plugin/isLeapYear';
import isoWeek from 'dayjs/plugin/isoWeek';
import isoWeeksInYear from 'dayjs/plugin/isoWeeksInYear';
import isFunction from 'lodash/isFunction';
import * as React from 'react';
import { ConfigContext } from '../config-provider';
import { PickerColumn, PickerColumnItem } from '../picker-view';
import { ColumnType, COLUMN_TYPE, DatePickerFilter, MERIDIEM, RenderLabel } from './interface';

dayjs.extend(isoWeek);
dayjs.extend(isoWeeksInYear);
dayjs.extend(isLeapYear);

export const generateMeridiemHour = (hour: number, hasMeridiem: boolean) => {
  if (!hasMeridiem) return hour;
  if (hour === 0) hour = 12;
  if (hour > 12) hour -= 12;
  return hour;
};

export const useRenderLabel = (renderLabel?: RenderLabel): RenderLabel => {
  const { locale: globalLocale } = React.useContext(ConfigContext);
  const locale = globalLocale?.DatePickerView;

  return React.useCallback(
    (type, value) => {
      if (isFunction(renderLabel)) {
        return renderLabel(type, value);
      }

      switch (type) {
        case COLUMN_TYPE.YEAR:
          return value + locale.year;
        case COLUMN_TYPE.MONTH:
          return value + locale.month;
        case COLUMN_TYPE.WEEK:
          return value + locale.week;
        case COLUMN_TYPE.WEEK_DAY:
          return locale.weeks[value - 1];
        case COLUMN_TYPE.DAY:
          return value + locale.day;
        case COLUMN_TYPE.HOUR:
          return value + locale.hour;
        case COLUMN_TYPE.MINUTE:
          return value + locale.minute;
        case COLUMN_TYPE.SECOND:
          return value + locale.second;
        case COLUMN_TYPE.MERIDIEM:
          return value === MERIDIEM.AM ? locale.am : locale.pm;
        default:
          return value.toString();
      }
    },
    [renderLabel],
  );
};

export const padZero = (value: string | number, length = 2) => {
  return String(value).padStart(length, '0');
};

export const generateDatePickerColumns = (
  selectedValue: Date,
  min: Date,
  max: Date,
  columnType: ColumnType[],
  renderLabel: RenderLabel,
  filter?: DatePickerFilter,
) => {
  const hasMeridiem = columnType.includes(COLUMN_TYPE.MERIDIEM);

  const date = dayjs(selectedValue);
  const minDate = dayjs(min);
  const maxDate = dayjs(max);

  const minYear = min.getFullYear();
  const minMonth = min.getMonth() + 1;
  const minDay = min.getDate();
  const minWeek = minDate.isoWeek();
  const minWeekday = minDate.isoWeekday();
  const minHour = min.getHours();
  const minMinute = min.getMinutes();
  const minSecond = min.getSeconds();

  const maxYear = max.getFullYear();
  const maxMonth = max.getMonth() + 1;
  const maxDay = max.getDate();
  const maxWeek = maxDate.isoWeek();
  const maxWeekday = maxDate.isoWeekday();
  const maxHour = max.getHours();
  const maxMinute = max.getMinutes();
  const maxSecond = max.getSeconds();

  const selectedYear = date.year();
  const selectedMonth = date.month() + 1;
  const selectedDay = date.date();
  const selectedWeek = date.isoWeek();
  const selectedHour = date.hour();
  const selectedMinute = date.minute();
  const selectedYearWeeks = dayjs(`${selectedYear}/01/01`).isoWeeksInYear();

  const isInMinYear = selectedYear === minYear;
  const isInMaxYear = selectedYear === maxYear;
  const isInMinMonth = isInMinYear && selectedMonth === minMonth;
  const isInMaxMonth = isInMaxYear && selectedMonth === maxMonth;
  const isInMinDay = isInMinMonth && selectedDay === minDay;
  const isInMaxDay = isInMaxMonth && selectedDay === maxDay;
  const isInMinWeek = isInMinYear && selectedWeek === minWeek;
  const isInMaxWeek = isInMaxYear && selectedWeek === maxWeek;
  const isInMinHour = isInMinDay && selectedHour === minHour;
  const isInMaxHour = isInMaxDay && selectedHour === maxHour;
  const isInMinMinute = isInMinHour && selectedMinute === minMinute;
  const isInMaxMinute = isInMaxHour && selectedMinute === maxMinute;

  const firstDayInSelectedMonth = dayjs(date).startOf('month');

  const generateColumns = <T extends PickerColumnItem>(
    from: number,
    to: number,
    type: ColumnType,
    iteratee: (index: number) => T,
  ) => {
    let options: T[] = [];
    for (let i = from; i <= to; i++) {
      options.push(iteratee(i));
    }

    if (isFunction(filter)) {
      options = options.filter((option) =>
        filter(type, {
          get value() {
            return option.value as number;
          },
          get date() {
            return selectedValue;
          },
        }),
      );
    }

    return options;
  };

  const generateMeridiemHourColumn = (hour: number) => {
    if (!hasMeridiem) return hour;
    if (hour > 12) hour -= 12;
    return hour;
  };

  const columns: PickerColumn = [];

  columnType?.forEach((type) => {
    switch (type) {
      case COLUMN_TYPE.YEAR:
        columns.push(
          generateColumns(minYear, maxYear, COLUMN_TYPE.YEAR, (index) => ({
            value: index,
            label: renderLabel(COLUMN_TYPE.YEAR, index),
          })),
        );
        break;
      case COLUMN_TYPE.MONTH:
        columns.push(
          generateColumns(
            isInMinYear ? minMonth : 1,
            isInMaxYear ? maxMonth : 12,
            COLUMN_TYPE.MONTH,
            (index) => ({
              value: index,
              label: renderLabel(COLUMN_TYPE.MONTH, index),
            }),
          ),
        );
        break;
      case COLUMN_TYPE.WEEK:
        columns.push(
          generateColumns(
            isInMinYear ? minWeek : 1,
            isInMaxYear ? maxWeek : selectedYearWeeks,
            COLUMN_TYPE.WEEK,
            (index) => ({
              value: index,
              label: renderLabel(COLUMN_TYPE.WEEK, index),
            }),
          ),
        );
        break;
      case COLUMN_TYPE.WEEK_DAY:
        columns.push(
          generateColumns(
            isInMinWeek ? minWeekday : 1,
            isInMaxWeek ? maxWeekday : 7,
            COLUMN_TYPE.WEEK_DAY,
            (index) => ({
              value: index,
              label: renderLabel(COLUMN_TYPE.WEEK_DAY, index),
            }),
          ),
        );
        break;
      case COLUMN_TYPE.DAY:
        columns.push(
          generateColumns(
            isInMinMonth ? minDay : 1,
            isInMaxMonth ? maxDay : firstDayInSelectedMonth.daysInMonth(),
            COLUMN_TYPE.DAY,
            (index) => ({
              value: index,
              label: renderLabel(COLUMN_TYPE.DAY, index),
            }),
          ),
        );
        break;
      case COLUMN_TYPE.MERIDIEM:
        columns.push(
          generateColumns(MERIDIEM.AM, MERIDIEM.PM, COLUMN_TYPE.MERIDIEM, (index) => ({
            value: index,
            label: renderLabel(COLUMN_TYPE.MERIDIEM, index),
          })),
        );
        break;
      case COLUMN_TYPE.HOUR:
        columns.push(
          generateColumns(
            generateMeridiemHourColumn(isInMinDay ? minHour : 0),
            generateMeridiemHourColumn(isInMaxDay ? maxHour : 23),
            COLUMN_TYPE.HOUR,
            (index) => ({
              value: index,
              label: renderLabel(COLUMN_TYPE.HOUR, generateMeridiemHour(index, hasMeridiem)),
            }),
          ),
        );
        break;
      case COLUMN_TYPE.MINUTE:
        columns.push(
          generateColumns(
            isInMinHour ? minMinute : 0,
            isInMaxHour ? maxMinute : 59,
            COLUMN_TYPE.MINUTE,
            (index) => ({
              value: index,
              label: renderLabel(COLUMN_TYPE.MINUTE, index),
            }),
          ),
        );
        break;
      case COLUMN_TYPE.SECOND:
        columns.push(
          generateColumns(
            isInMinMinute ? minSecond : 0,
            isInMaxMinute ? maxSecond : 59,
            COLUMN_TYPE.SECOND,
            (index) => ({
              value: index,
              label: renderLabel(COLUMN_TYPE.SECOND, index),
            }),
          ),
        );
        break;
    }
  });

  return columns;
};

export const dateToNumberArray = (value: Date, columnType: ColumnType[]) => {
  const date = dayjs(value);
  const hasWeek = columnType?.includes(COLUMN_TYPE.WEEK);
  const hasMeridiem = columnType?.includes(COLUMN_TYPE.MERIDIEM);
  return columnType?.map((type) => {
    switch (type) {
      case COLUMN_TYPE.YEAR:
        return hasWeek ? date.isoWeekYear() : date.year();
      case COLUMN_TYPE.MONTH:
        return date.month() + 1;
      case COLUMN_TYPE.DAY:
        return date.date();
      case COLUMN_TYPE.WEEK:
        return date.isoWeek();
      case COLUMN_TYPE.WEEK_DAY:
        return date.isoWeekday();
      case COLUMN_TYPE.MERIDIEM:
        return date.hour() >= 12 ? MERIDIEM.PM : MERIDIEM.AM;
      case COLUMN_TYPE.HOUR:
        return generateMeridiemHour(date.hour(), hasMeridiem);
      case COLUMN_TYPE.MINUTE:
        return date.minute();
      case COLUMN_TYPE.SECOND:
        return date.second();
      default:
        return undefined;
    }
  });
};

export const numberArrayToDate = (
  currentValue: Date,
  changedPickerValue: number[],
  columnType: ColumnType[],
  level: number,
) => {
  const date = dayjs(currentValue);
  const types = Object.fromEntries(
    columnType.map((type, index) => [type, changedPickerValue[index]]),
  );
  const currentColumnType = columnType[level];

  const getMeridiemHour = (meridiem: number, hour: number) => {
    if (meridiem === MERIDIEM.AM && hour >= 12) {
      hour -= 12;
    }
    if (meridiem === MERIDIEM.PM && hour < 12) {
      hour += 12;
    }
    return hour;
  };

  const getHour = (hour: number) => {
    const meridiemIndex = columnType?.findIndex((i) => i === COLUMN_TYPE.MERIDIEM);
    const meridiemValue = changedPickerValue[meridiemIndex] as number;
    return getMeridiemHour(meridiemValue, hour);
  };

  const year = types.year ?? date.isoWeekYear() ?? 1900;
  const month = types.month ?? date.month() + 1 ?? 0;
  const week = types.week ?? 1;
  const weekDay = types['week-day'] ?? 1;
  const day = types.day ?? date.date() ?? 1;
  const hour = types.hour ?? 0;
  const minute = types.minute ?? 0;
  const second = types.second ?? 0;

  const selectedMonth = month - 1;
  const selectedDay = Math.min(day, dayjs(new Date(year, selectedMonth, 1)).daysInMonth());
  const selectedHour = getHour(hour);
  const selectedDate = dayjs().hour(selectedHour).minute(minute).second(second);

  if (columnType.includes(COLUMN_TYPE.WEEK)) {
    if (currentColumnType === COLUMN_TYPE.MONTH) {
      return selectedDate.year(year).month(selectedMonth).toDate();
    }
    if (currentColumnType === COLUMN_TYPE.DAY) {
      return selectedDate.year(year).isoWeek(week).date(selectedDay).toDate();
    }
    return selectedDate.year(year).isoWeek(week).isoWeekday(weekDay).toDate();
  }

  return selectedDate.year(year).month(selectedMonth).date(selectedDay).toDate();
};
