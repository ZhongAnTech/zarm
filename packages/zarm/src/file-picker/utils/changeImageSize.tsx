/**
 * change img size or quality todo change size
 *
 * @param img
 * @param quality
 * @param fileType
 */
export default function changeImageSize(
  img: HTMLImageElement,
  quality: number | undefined,
  fileType: string,
) {
  const canvas = document.createElement('canvas');
  const ctx = canvas.getContext('2d');

  if (!ctx) {
    console.error('请使用高版本浏览器，该版本浏览器不支持生成缩略图');
    return '';
  }

  // const MAX_WIDTH = 800;
  // const MAX_HEIGHT = 600;
  const { width, height } = img;

  // if (width > height) {
  //   if (width > MAX_WIDTH) {
  //     height *= MAX_WIDTH / width;
  //     width = MAX_WIDTH;
  //   }
  // } else if (height > MAX_HEIGHT) {
  //   width *= MAX_HEIGHT / height;
  //   height = MAX_HEIGHT;
  // }

  canvas.width = width;
  canvas.height = height;

  // 在canvas绘制前填充白色背景
  ctx.fillStyle = '#fff';
  ctx.fillRect(0, 0, canvas.width, canvas.height);

  ctx.drawImage(img, 0, 0, width, height);

  // 是否有效的压缩比例（0 - 1）
  if (quality && !(quality > 0 && quality <= 1)) {
    console.error('请输入有效的压缩比例, 没有将默认使用 0.92');
  }

  return canvas.toDataURL(fileType, quality);
}
