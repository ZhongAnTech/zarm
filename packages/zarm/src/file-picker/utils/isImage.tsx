/**
 * detect the file is image
 *
 * @param fileType
 */
export default function isImage(fileType: string): boolean {
  const imageType = /image.*/;

  return !!fileType.match(imageType);
}
