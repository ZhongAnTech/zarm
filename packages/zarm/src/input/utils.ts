export function getValue(props, defaultValue: string | number = ''): string | number {
  if (typeof props.value !== 'undefined' && props.value !== null) {
    return props.value;
  }
  if (typeof props.defaultValue !== 'undefined' && props.defaultValue !== null) {
    return props.defaultValue;
  }
  return defaultValue;
}
