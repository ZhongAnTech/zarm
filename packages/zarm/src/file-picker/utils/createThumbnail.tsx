import changeImageSize from './changeImageSize';

export interface IFileProps {
  file: File;
  quality?: number;
  fileType: string;
  maxWidth?: number | string;
  maxHeight?: number | string;
}

/**
 * generate preview image
 * modify image quality
 *
 * @param file
 * @param quality
 * @param fileType
 * @param maxWidth
 * @param maxHeight
 * @param callback
 */
export default function createThumbnail({
  file,
  quality,
  fileType,
  maxWidth,
  maxHeight,
}: IFileProps) {
  return new Promise<string>((resolve) => {
    const img = document.createElement('img');

    window.URL = window.URL || window.webkitURL;

    img.onload = () => {
      let imgUrl: string;

      if (quality || maxWidth || maxHeight) {
        imgUrl = changeImageSize(img, quality, fileType);
      } else {
        imgUrl = img.src;
      }

      resolve(imgUrl || '');
    };

    img.src = window.URL.createObjectURL(file);
  });
}

// FileReader
// export default function createThumbnail({ file, quality, fileType, maxWidth, maxHeight }, callback) {
//   const img = document.createElement('img');
//   const reader = new FileReader();

//   reader.onload = (e) => {
//     img.src = e.target.result;

//     callback(img.src);
//   };

//   reader.readAsDataURL(file);
// }
