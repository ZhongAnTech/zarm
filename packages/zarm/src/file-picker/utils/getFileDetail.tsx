import isImage from './isImage';

/**
 * get file detail
 *
 * @param file
 */
export default function getFileDetail(file: File) {
  const fileSize = file.size;
  const fileName = file.name;
  const fileType = file.type || fileName.substr(fileName.lastIndexOf('.') + 1);
  const isPic = isImage(fileType);

  return {
    fileName,
    fileSize,
    fileType,
    isPic,
  };
}
