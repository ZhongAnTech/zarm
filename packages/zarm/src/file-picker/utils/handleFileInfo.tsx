import getFileDetail from './getFileDetail';
import createThumbnail from './createThumbnail';
import type { FileObject } from '../interface';

/**
 * handle single file and get file info
 *
 * @param file
 * @param quality
 * @param callback
 */
export default function handleFileInfo(
  { file, quality }: { file: File; quality?: number },
  callback: (data: FileObject) => void,
): void {
  const { fileSize, fileType, isPic, fileName } = getFileDetail(file);

  const fileDetail: FileObject = {
    file,
    fileType,
    fileSize,
    fileName,
    thumbnail: '',
  };

  const callbackFunc = (url: string) => {
    fileDetail.thumbnail = url;

    callback(fileDetail);
  };

  if (isPic) {
    createThumbnail({
      file,
      quality,
      fileType,
    })
      .then((url) => callbackFunc(url))
      .catch((e) => console.error(e));
  } else {
    callback(fileDetail);
  }
}
