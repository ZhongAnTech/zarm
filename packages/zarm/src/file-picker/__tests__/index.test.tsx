import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { mocked } from 'ts-jest/utils';
import FilePicker from '../index';
import type { FileObject } from '../interface';
import { mockCreateObjectURL, mockResetCreateObjectURL } from '../../../tests/utils';
import handleFileInfo from '../utils/handleFileInfo';

jest.mock('../utils/handleFileInfo');

const mHandleFileInfo = mocked(handleFileInfo);

describe('file picker', () => {
  it('render correctly', () => {
    const props = {
      onChange: jest.fn(),
      children: <button>add</button>,
    };
    const wrapper = render(<FilePicker {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe('file picker event', () => {
  const props = {
    accept: 'image/jpg, image/jpeg, image/gif, image/png',
    onChange: jest.fn(),
    quality: 0.3,
    children: <button>add</button>,
  };
  const mCreateObjectURL = jest.fn();
  const file = new File([''], './test.jpg', { type: 'image/jpeg' });
  const mFileDetail = {
    file,
    fileType: file.type,
    fileName: file.name,
    fileSize: file.size,
    thumbnail: 'data:,',
  };

  beforeAll(() => {
    mockCreateObjectURL(mCreateObjectURL);
  });
  afterAll(() => {
    mockResetCreateObjectURL();
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('select image & onBeforeSelect is not func', () => {
    const _props = {
      ...props,
      multiple: true,
      onBeforeSelect: false as any,
    };
    mHandleFileInfo.mockImplementation((_, callback: (data: FileObject) => void) => {
      callback(mFileDetail);
    });
    const wrapper = mount(<FilePicker {..._props} />);

    const mClickEvent = { preventDefault: jest.fn() };
    wrapper.find('.za-file-picker__input').simulate('click', mClickEvent);
    expect(mClickEvent.preventDefault).not.toBeCalled();

    wrapper.find('.za-file-picker__input').simulate('change', {
      target: { files: [file] },
    });
    expect(props.onChange).toBeCalledWith([mFileDetail]);
    expect(props.onChange).toBeCalledTimes(1);
  });

  it('should handle change event if files is an array of File instance', () => {
    let getFileInfo: (data: FileObject) => void;
    mHandleFileInfo.mockImplementation((_, callback: (data: FileObject) => void) => {
      getFileInfo = callback;
      callback(mFileDetail);
    });
    const wrapper = mount(<FilePicker {...props} />);
    wrapper.find('.za-file-picker__input').simulate('change', {
      target: { files: [file] },
    });

    expect(mHandleFileInfo).toBeCalledWith({ file, quality: 0.3 }, getFileInfo!);
    expect(props.onChange).toBeCalledWith(mFileDetail);
  });

  it('click children', () => {
    const wrapper = mount(<FilePicker {...props} />);

    wrapper.find('button').simulate('click');
    wrapper.find('.za-file-picker__input').simulate('change', {
      target: { files: [file] },
    });
    expect(props.onChange).toBeCalled();
  });
});

describe('file picker disabled', () => {
  const props = {
    accept: 'image/jpg, image/jpeg, image/gif, image/png',
    onChange: jest.fn(),
    onBeforeSelect: jest.fn(),
    children: <button>foo</button>,
  };

  it('on before select return false', () => {
    const _props = {
      ...props,
      onBeforeSelect: jest.fn(() => false),
    };

    const wrapper = mount(<FilePicker {..._props} />);

    wrapper.find('.za-file-picker__input').simulate('click');
    expect(props.onChange).not.toHaveBeenCalled();
  });

  it('disabled', () => {
    const _props = {
      ...props,
      disabled: true,
    };

    const wrapper = mount(<FilePicker {..._props} />);

    wrapper.find('.za-file-picker__input').simulate('click');
    expect(props.onChange).not.toHaveBeenCalled();
  });
});
