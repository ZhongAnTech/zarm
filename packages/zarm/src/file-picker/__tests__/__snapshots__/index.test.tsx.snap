// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`file picker render correctly 1`] = `
<div
  class="za-file-picker"
>
  <input
    class="za-file-picker__input"
    type="file"
  />
  <button
    class="needsclick"
  >
    add
  </button>
</div>
`;
